commit 8052f141388fd94a608f719f2205c031d7c94361 (refs/changes/83/2183/2)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2017-07-17T17:15:48+02:00 (2 years, 3 months ago)
    
    TOOLS-1772 Add "--just-update-symlink" option to `sdcadm experimental update-agents`
    Reviewed by: Trent Mick <trentm@gmail.com>
    Approved by: Trent Mick <trentm@gmail.com>

diff --git a/lib/cli/do_update_agents.js b/lib/cli/do_update_agents.js
index 70acb87..7dd3121 100644
--- a/lib/cli/do_update_agents.js
+++ b/lib/cli/do_update_agents.js
@@ -53,6 +53,8 @@ function updateAgents(options, callback) {
     assert.optionalBool(options.justDownload, 'options.justDownload');
     assert.optionalBool(options.skipLatestSymlink,
         'options.skipLatestSymlink');
+    assert.optionalBool(options.justUpdateSymlink,
+        'options.justUpdateSymlink');
     assert.optionalBool(options.yes, 'options.yes');
     assert.optionalBool(options.all, 'options.all');
     assert.optionalArrayOfString(options.servers, 'options.servers');
@@ -60,7 +62,8 @@ function updateAgents(options, callback) {
     assert.func(callback, 'callback');
 
     if ((options.all && options.servers) ||
-        (!options.all && !options.servers)) {
+        (!options.all && !options.servers &&
+         !options.justDownload && !options.justUpdateSymlink)) {
         callback(new errors.UsageError(
             'must specify exactly one of "options.all" or "options.servers"'));
         return;
@@ -77,6 +80,7 @@ function updateAgents(options, callback) {
     var progress = options.progress;
     var justDownload = options.justDownload;
     var skipLatestSymlink = options.skipLatestSymlink;
+    var justUpdateSymlink = options.justUpdateSymlink;
     var hist;
 
     function setImageToLatest(cb) {
@@ -136,8 +140,9 @@ function updateAgents(options, callback) {
          * it:
          */
         function urDiscoveryGetReady(ctx, next) {
-            if (justDownload) {
-                return next();
+            if (justDownload || justUpdateSymlink) {
+                next();
+                return;
             }
 
             sdcadm.getUrConnection(function (err, urconn) {
@@ -185,6 +190,19 @@ function updateAgents(options, callback) {
             }
         },
 
+        function verifyFilepath(ctx, next) {
+            if (!justUpdateSymlink) {
+                next();
+                return;
+            }
+            if (!filepath) {
+                next(new Error('existing file must be specified when using ' +
+                    '"--just-update-symlink" option'));
+                return;
+            }
+            next();
+        },
+
         /*
          * If we are about to download an image, first check to see if that
          * image is already available locally (verifying via checksum).
@@ -257,14 +275,16 @@ function updateAgents(options, callback) {
         },
 
         function listServers(ctx, next) {
-            if (justDownload) {
-                return next();
+            if (justDownload || justUpdateSymlink) {
+                next();
+                return;
             }
             progress('Finding servers to update');
             // Get all servers to validate if unsetup servers are selected.
             sdcadm.cnapi.listServers({}, function (err, servers) {
                 if (err) {
-                    return next(err);
+                    next(err);
+                    return;
                 }
                 ctx.allServers = servers;
                 next();
@@ -272,8 +292,9 @@ function updateAgents(options, callback) {
         },
 
         function findServersToUpdate(ctx, next) {
-            if (justDownload) {
-                return next();
+            if (justDownload || justUpdateSymlink) {
+                next();
+                return;
             }
 
             if (options.all) {
@@ -328,8 +349,9 @@ function updateAgents(options, callback) {
         },
 
         function urDiscovery(ctx, next) {
-            if (justDownload) {
-                return next();
+            if (justDownload || justUpdateSymlink) {
+                next();
+                return;
             }
 
             common.urDiscovery({
@@ -340,10 +362,11 @@ function updateAgents(options, callback) {
                 urconn: ctx.urconn
             }, function (err, urAvailServers) {
                 if (err) {
-                    return next(err);
+                    next(err);
+                    return;
                 }
                 ctx.urServersToUpdate = urAvailServers;
-                return next();
+                next();
             });
         },
 
@@ -365,32 +388,40 @@ function updateAgents(options, callback) {
                     'Download agentsshar %s\n    (%s)',
                     image.uuid, image.version)));
             }
-            if (!justDownload) {
+            if (!justDownload && !justUpdateSymlink) {
                 progress(common.indent(format(
                     'Update GZ agents on %d (of %d) servers using\n' +
                     '    agentsshar %s', ctx.serversToUpdate.length,
                     ctx.allServers.length,
                     (filepath ? filepath : image.version))));
             }
+            if (justUpdateSymlink && filepath) {
+                progress(common.indent(format('Update ' +
+                    '\'/usbkey/extra/agents/latest\' symkink to %s',
+                    filepath)));
+            }
             progress('');
             if (options.yes) {
-                return next();
+                next();
+                return;
             }
             var msg = 'Would you like to continue? [y/N] ';
             common.promptYesNo({msg: msg, default: 'n'}, function (answer) {
                 if (answer !== 'y') {
                     progress('Aborting agents update');
-                    return callback();
+                    callback();
+                    return;
                 }
                 progress('');
                 startTime = Date.now(); // Reset to not count confirm time.
-                return next();
+                next();
             });
         },
 
         function saveChangesToHistory(_, next) {
-            if (justDownload) {
-                return next();
+            if (justDownload || justUpdateSymlink) {
+                next();
+                return;
             }
             var change = {
                 service: {
@@ -404,10 +435,11 @@ function updateAgents(options, callback) {
                 changes: [change]
             }, function (err, hst) {
                 if (err) {
-                    return next(err);
+                    next(err);
+                    return;
                 }
                 hist = hst;
-                return next();
+                next();
             });
         },
 
@@ -498,8 +530,9 @@ function updateAgents(options, callback) {
         },
 
         function updateCNAgents(ctx, next) {
-            if (justDownload) {
-                return next();
+            if (justDownload || justUpdateSymlink) {
+                next();
+                return;
             }
 
             progress('Starting agentsshar update on %d servers',
@@ -684,8 +717,9 @@ function updateAgents(options, callback) {
         },
 
         function doCleanup(ctx, next) {
-            if (justDownload) {
-                return next();
+            if (justDownload || justUpdateSymlink) {
+                next();
+                return;
             }
             if (ctx.deleteAgentssharOnFinish) {
                 progress('Deleting temporary %s', filepath);
@@ -705,8 +739,9 @@ function updateAgents(options, callback) {
         // instances and until we move from update agents using a shar file
         // to the individual agents update.
         function refreshSysinfo(ctx, next) {
-            if (justDownload) {
-                return next();
+            if (justDownload || justUpdateSymlink) {
+                next();
+                return;
             }
 
             progress('Reloading sysinfo on updated servers');
@@ -763,8 +798,9 @@ function updateAgents(options, callback) {
          * race.
          */
         function refreshConfigAgents(ctx, next) {
-            if (justDownload) {
-                return next();
+            if (justDownload || justUpdateSymlink) {
+                next();
+                return;
             }
             progress('Refreshing config-agent on all the updated servers');
 
@@ -793,8 +829,9 @@ function updateAgents(options, callback) {
         if (err === true) { // early abort signal
             err = null;
         }
-        if (justDownload) {
-            return callback(err);
+        if (justDownload || justUpdateSymlink) {
+            callback(err);
+            return;
         }
 
         if (!hist) {
@@ -803,7 +840,8 @@ function updateAgents(options, callback) {
             hist.error = err;
             sdcadm.history.updateHistory(hist, function (histErr) {
                 if (err) {
-                    return callback(err);
+                    callback(err);
+                    return;
                 }
                 progress('Successfully updated agents (%s)',
                     common.humanDurationFromMs(Date.now() - startTime));
@@ -839,7 +877,8 @@ function do_update_agents(subcmd, opts, args, cb) {
             'cannot specify "--all" and explicit servers: ' +
             servers.join(' ')));
         return;
-    } else if (!opts.all && !servers && !opts.just_download) {
+    } else if (!opts.all && !servers && !opts.just_download &&
+        !opts.just_update_symlink) {
         cb(new errors.UsageError(
             'either --all option or explicitly specifying ' +
             'SERVER(s) is required'));
@@ -858,6 +897,7 @@ function do_update_agents(subcmd, opts, args, cb) {
         progress: self.progress,
         justDownload: opts.just_download,
         skipLatestSymlink: opts.skip_latest_symlink,
+        justUpdateSymlink: opts.just_update_symlink,
         yes: opts.yes,
         servers: servers,
         all: opts.all,
@@ -923,6 +963,12 @@ do_update_agents.options = [
         help: 'Do not modify the file pointed at by ' +
             '\'/usbkey/extra/agents/latest\' symlink.'
     },
+    {
+        names: ['just-update-symlink'],
+        type: 'bool',
+        help: 'Only update \'/usbkey/extra/agents/latest\' ' +
+            'to the given file path.'
+    },
     {
         names: ['all', 'a'],
         type: 'bool',
